text/0000-succinct-atomic-swaps.md

```diff
@@ -0,0 +1,174 @@
+
+Without automated defenses, a malicious party could trick a user into deriving the same nonce (by supplying a duplicate atomic ID). The malicious user could then recover funds on the other chain without completing the atomic swap protocol.
+
+Regular blinding factors and signature nonces use a strictly increasing counter to prevent deriving the same factors more than once. However, this technique alone won't work for atomic nonces. The atomic ID is shared between parties, meaning a malicious attacker could attempt to trick a user into reusing an atomic ID.
+
+Currently, the automated defense is to use a space-efficient filter (`scalable cuckoo filter`), which is stored in the wallet's database. The filter cannot be reconstructed if the wallet file is deleted and recovered, so it is recommended to use a new keymask for any future atomic swaps.
+
+## Adaptor signatures
+[adaptor-signatures]: #adaptor-signatures
+
+Adaptor signatures were originally invented by Andrew Poelstra [\[3\]](#references). Adaptor signatures use additive properties of Schnorr signatures to commit to a secret, then later reveal it by creating a full Schnorr signature over the same message.
+
+
+## Protocol rounds
+[protocol-rounds]: #protocol-rounds
+
+In Grin, let `kB` be Bob's random kernel nonce, `rB` be Bob's random blinding factor, `nB` be Bob's atomic nonce. Let `kA` and `rA` be Alice's random kernel nonce and blinding factor.
+
+Alice and Bob must perform an atomic swap refund transaction before the main transaction. After a Alice receives the slate in the fourth refund round, both parties can lock funds on the other chain.
+
+An illustration of the protocol (credit @tromp + @RubenSomsen):
+
+```
++----------+        +-------------+  1 day   +-------------+  2 days     +----------+
+| 87k Grin |  --->  | 87k Grin    |  ----->  | 87k Grin    |  -------->  | 87k Grin |
+| Alice    |        | Alice + Bob |          | Alice + Bob |  reveals    | Alice    |
++----------+        +-------------+          +-------------+  secretA    +----------+
+                                                       
+                            |                       | 3
+                            |                       | days
+                            |                       v
+                            |                         
+                            |                +----------+
+                            +------------->  | 87k Grin |
+                             reveals         | Bob      |
+                             secretB         +----------+
+
+
+
++-------+         +-------------------+
+| 1 BTC |  ---->  | 1 BTC             |
+| Bob   |         | secretA + secretB |
++-------+         +-------------------+
+```
+
+The following is a description of the main transaction rounds.
+
+1. `init_atomic_swap`
+
+In the first round, Alice sends her public random kernel nonce and blinding factor (`kA*G` and `rA*G`) to Bob.
+
```

Comment:
There seems to be a ton of details missing here. Where is the description of how to build and verify all 5 transactions shown in the diagram?

Reply:
> There seems to be a ton of details missing here. Where is the description of how to build and verify all 5 transactions shown in the diagram?

There are only three transactions? I took the diagram you provided to be showing the state transitions, not independent transactions.

I can add a few lines about how transactions are built, though. Basically, the transactions (refund and main) on the Grin side are built like normal 1-in-2-out transactions. The refund is a self-spend co-signed by the other party (they have no inputs/outputs). The main transaction is essentially just a 1-in-2-out transaction, that reveals Alice's secret.

That hardly seems like "a ton of details"...

Reply:
> There seems to be a ton of details missing here. Where is the description of how to build and verify all 5 transactions shown in the diagram?

There are only three transactions? I took the diagram you provided to be showing the state transitions, not independent transactions.

I can add a few lines about how transactions are built, though. Basically, the transactions (refund and main) on the Grin side are built like normal 1-in-2-out transactions. The refund is a self-spend co-signed by the other party (they have no inputs/outputs). The main transaction is essentially just a 1-in-2-out transaction, that reveals Alice's secret.

The non-Grin transaction uses a multisignature (potentially time-locked). If a time-lock is used, the funds are returned to the funder when the timer expires.

That seems around a thousand kilograms short of a "ton"

Reply:
> There seems to be a ton of details missing here. Where is the description of how to build and verify all 5 transactions shown in the diagram?

There are only three transactions in the swap proper (plus the two funding transactions). I was only describing the transactions actually involved in the funded swap, but can add a sentence describing the need to fund the swap.  I took the diagram you provided to be showing the state transitions, not independent transactions.

I can also add a few lines about how transactions are built. Basically, the transactions (refund and main) on the Grin side are built like normal 1-in-2-out transactions. The refund is a self-spend co-signed by the other party (they have no inputs/outputs from the other party). The main transaction is essentially just a 1-in-2-out transaction, that reveals Alice's secret.

The non-Grin transaction uses a multisignature (potentially time-locked). If a time-lock is used, the funds are returned to the funder when the timer expires.

Reply:
My original picture showed outputs in boxes and transactions (with locktimes and possible reveal properties) as edges between them.
On the Grin side, there were 5 outputs, and 5 transactions between them.
In your new picture you seem to have put the transaction names on the boxes, which makes no sense to me.

Reply:
Like you said, you cannot implement the 2-tx swap without relative timelocks.
What you implement is the 3-tx swap, where the 2 Grin txs of a succesful exectution are Alice's funding tx, and what Somsen calls the Success transaction.

What Somsen calls the Revoke transaction, Refund transaction #1, and Timeout transaction
are needed for handling abnormal executions.

Your new ASCII diagram shows amount and blinding factors (output properties) in boxes,
and shows locktimes and revealing properties (kernel properties) on edges. Yet it names boxes as transactions. This is quite confusing.

Reply:
> What Somsen calls the Revoke transaction, Refund transaction #1, and Timeout transaction are needed for handling abnormal executions.

Right, and I did not include the Revoke transaction, because it only makes sense as a setup for the two relative timelocked transactions (`Refund #2` and `Timeout`). If you switch the transactions to use absolute timelocks, they effectively serve the same purpose of the `Refund #1` transaction, so why bloat/complicate the impl? You gain no advantage, since the `Revoke` transaction flow is meant to be an automated way to abort the swap, or award Bob the coins if Alice is unresponsive/inactive.

> The relative timelocks are only needed in the 2-tx version

No, if you look at the [diagram](https://gist.githubusercontent.com/RubenSomsen/8853a66a64825716f51b409be528355f/raw/c4cf1e7b66e6d1aa775cabcae6464ddd722fc506/SuccinctAtomicSwap.svg), relative timelocks are needed for both the `Refund #2` and `Timeout` transactions.

> What you implement is the 3-tx swap, where the 2 Grin txs of a successful execution are Alice's funding tx, and what Somsen calls the Success transaction.

Again, no. I implemented what Somsen calls the `Success` and `Refund` transactions, and leave the `Revoke` (and its children) for future work when NRD kernels (i.e. relative timelocks) become available.

> Your new ASCII diagram shows amount and blinding factors (output properties) in boxes,
and shows locktimes and revealing properties (kernel properties) on edges. Yet it names boxes as transactions. This is quite confusing.

I don't understand what's confusing about the new diagram. To me it's a more straightforward representation of the version of atomic swaps that I've implemented. I'd be interested to get feedback from more people.

Reply:
> What Somsen calls the Revoke transaction, Refund transaction #1, and Timeout transaction are needed for handling abnormal executions.

Right, and I did not include the Revoke transaction, because it only makes sense as a setup for the two relative timelocked transactions (`Refund #2` and `Timeout`). If you switch the transactions to use absolute timelocks, they effectively serve the same purpose of the `Refund #1` transaction, so why bloat/complicate the impl? You gain no advantage, since the `Revoke` transaction flow is meant to be an automated way to abort the swap, or award Bob the coins if Alice is unresponsive/inactive.

> The relative timelocks are only needed in the 2-tx version

No, if you look at the [diagram](https://gist.githubusercontent.com/RubenSomsen/8853a66a64825716f51b409be528355f/raw/c4cf1e7b66e6d1aa775cabcae6464ddd722fc506/SuccinctAtomicSwap.svg), relative timelocks are needed for both the `Refund #2` and `Timeout` transactions.

> What you implement is the 3-tx swap, where the 2 Grin txs of a successful execution are Alice's funding tx, and what Somsen calls the Success transaction.

Again, no. I implemented what Somsen calls the `Success` and `Refund` transactions, and leave the `Revoke` (and its children) for future work when NRD kernels (i.e. relative timelocks) become available.

> Your new ASCII diagram shows amount and blinding factors (output properties) in boxes,
and shows locktimes and revealing properties (kernel properties) on edges. Yet it names boxes as transactions. This is quite confusing.

I don't understand what's confusing about the new diagram. To me it's a more straightforward representation of the version of atomic swaps that I've implemented. I'd be interested to get feedback from more people.

The new diagram is also closer to [Somsen's diagram](https://gist.githubusercontent.com/RubenSomsen/8853a66a64825716f51b409be528355f/raw/c4cf1e7b66e6d1aa775cabcae6464ddd722fc506/SuccinctAtomicSwap.svg) where the boxes with the inputs and outputs are labeled with the transaction titles. Is it confusing to you because you originally thought of the edges as transactions?

Reply:
> and leave the Revoke (and its children) for future work

I'm afraid you don't understand how SAS works. The Revoke tx is essential to its correct operation. The relative timelocks in the followups are not, and can safely be replaced by absolute timelocks if you always use the 3-tx version of SAS.

Reply:
> I'm afraid your understanding of SAS is not quite correct.

It's not that my understanding is incorrect, I am trying to implement SAS with what's currently available on Grin.

I tried to follow the SAS protocol as closely as possible, deferring to the implementation described in the official [Grin documentation](https://docs.grin.mw/wiki/transactions/contracts/#atomic-swap) for atomic swaps when something didn't seem possible.

Admittedly, my implementation is much closer to the Grin documentation atomic swaps than to SAS.

> The relative timelocks in the followups are not, and can safely be replaced by absolute timelocks if you always use the 3-tx version of SAS.

After re-reading the [SAS specification](https://gist.githubusercontent.com/RubenSomsen/8853a66a64825716f51b409be528355f/raw/c4cf1e7b66e6d1aa775cabcae6464ddd722fc506), I saw question in the FAQ about using absolute timelocks:

```
-  Is it possible to implement this protocol on chains which only support absolute timelocks?

   Yes, but then Bob must spend his swapped coins before the timelock expires (or use the 3 tx protocol). Be aware that the revoke_tx MUST confirm before the timeout_tx becomes valid, which may become a problem if fees suddenly rise. The refund_tx can also not be allowed to CPFP the timeout_tx, as they must confirm independently in order to invalidate the success_tx first.
```

I rewrote the RFC to more closely follow SAS, but have some questions about the feasibility on Grin.

- How to do the multisignature signing on `Grin Revoke`?
  - Like normal with Alice having full control over the UTXO(s) being spent?
  - Does it require Grin multisignatures to be implemented first?
- The `Grin Timeout` transaction sends Grin to Bob, without revealing Bob's atomic nonce to Alice. Alice cannot retrieve the funds locked on the other chain, if that happens.
  - Is this intentional, or an oversight in the protocol?
  - If intentional, should Bob have to reveal his atomic nonce to claim the `Grin Timeout` transaction?
  - If Bob doesn't reveal his nonce, this could be a Denial-of-Service attack vector on Alice
- It is unclear who the signer of `sigTimeoutAlice` and `sigTimeoutBob` are
  - Does Alice sign `sigTimeoutBob` and Bob sign `sigTimeoutAlice` (because `sigRefundBob` is signed by Alice)
  - If Alice signs `sigTimeoutBob`, she doesn't pre-sign the `Grin Timeout` transaction, so how does that work?